ivc/src/ivc/interpreter.rs

```diff
@@ -913,15 +916,101 @@ where
 {
 }
 
+/// Builds selectors for the IVC circuit.
+pub fn build_selectors<F, const N_COL_TOTAL: usize, const N_CHALS: usize>(
+    domain_size: usize,
+) -> Vec<Vec<F>>
+where
+    F: PrimeField,
+{
+    // 3*N + 6*N+2 + N+1 + 35*N + 5 + 1 + N_CHALS =
+    // 45N + 9 + N_CHALS
+    let mut selectors: Vec<Vec<F>> = vec![vec![F::zero(); domain_size]; N_BLOCKS];
+    let mut cur_row = 0;
+    for _i in 0..3 * N_COL_TOTAL {
+        selectors[0][cur_row] = F::one();
+        cur_row += 1;
+    }
+    for _i in 0..6 * N_COL_TOTAL + 2 {
+        selectors[1][cur_row] = F::one();
+        cur_row += 1;
+    }
+    for _i in 0..N_COL_TOTAL + 1 {
+        selectors[2][cur_row] = F::one();
+        cur_row += 1;
+    }
+    for _i in 0..(35 * N_COL_TOTAL + 5) {
+        selectors[3][cur_row] = F::one();
+        cur_row += 1;
+    }
+    for _i in 0..N_CHALS {
+        selectors[4][cur_row] = F::one();
+        cur_row += 1;
+    }
+    for _i in 0..1 {
+        selectors[5][cur_row] = F::one();
+        cur_row += 1;
+    }
+
+    selectors
+}
+
+/// This function generates constraitns for the whole IVC circuit.
+pub fn constrain_selectors<F, Env>(env: &mut Env)
+where
+    F: PrimeField,
+    Env: ColAccessCap<F, IVCColumn>,
+{
+    for i in 0..N_BLOCKS {
+        // Each selector must have value either 0 or 1.
+        let sel = env.read_column(IVCColumn::BlockSel(i));
+        env.assert_zero(sel.clone() * (sel.clone() - Env::constant(F::from(1u64))));
+    }
+}
+
+/// This function generates constraitns for the whole IVC circuit.
+pub fn constrain_ivc<F, Ff, Env>(env: &mut Env)
+where
+    F: PrimeField,
+    Ff: PrimeField,
+    Env: ColAccessCap<F, IVCColumn> + LookupCap<F, IVCColumn, IVCLookupTable<Ff>>,
+{
+    constrain_selectors(env);
+
+    let s0 = env.read_column(IVCColumn::BlockSel(0));
+    env.set_assert_mapper(Box::new(move |x| s0.clone() * x));
```

Comment:
I would appreciate some comments explaining what these constraints are doing because I cannot judge whether it looks okay to me or not.

Reply:
Will do!